### PR TITLE
[MOB-15384] AEPIdentity - Handle empty server url value for syncIdentifiers

### PR DESCRIPTION
**Pull request by addb**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: AEPIdentity/Sources/IdentityState.swift [hunk 1]
```diff
@@ -200,8 +201,10 @@ class IdentityState {
         let hasDpids = !(dpids?.isEmpty ?? true)
 
         if identityProperties.ecid != nil, !hasIds, !hasDpids, !needResync {
+            Log.trace(label: "\(LOG_TAG):\(#function)", "Not syncing identifiers, ECID found in persistence.")
             syncForIds = false
         } else if identityProperties.ecid == nil {
+            Log.trace(label: "\(LOG_TAG):\(#function)", "Generating new ECID, ECID not found in persistence.")
             identityProperties.ecid = ECID()
```
Comment (emdobrin) [inline AEPIdentity/Sources/IdentityState.swift:208]: Let's add a similar log when generating ECID inside handleNetworkResponse

File: AEPIdentity/Sources/Identity.swift [hunk 2]
```diff
@@ -81,6 +81,7 @@ import Foundation
             createSharedState(data: state.identityProperties.toEventData(), event: nil)
         }
 
+        Log.trace(label: "\(name):\(#function)", "Cannot process events yet. Identity is booting up.")
```
[the remark this answers is not in the document]
Comment (emdobrin) [reply AEPIdentity/Sources/Identity.swift:84]: Let's remove this as you added more logs inside

File: AEPIdentity/Sources/IdentityState.swift [hunk 3]
```diff
@@ -43,7 +43,10 @@ class IdentityState {
     /// - Returns: True if we should share state after bootup, false otherwise
     func bootupIfReady(configSharedState: [String: Any], event: Event) -> Bool {
         // Only bootup once we can perform the first force sync
-        guard readyForSyncIdentifiers(event: event, configurationSharedState: configSharedState) else { return false }
+        guard readyForSyncIdentifiers(event: event, configurationSharedState: configSharedState) else {
+            Log.trace(label: "\(LOG_TAG):\(#function)", "Cannot boot up, not ready for syncIdentifiers.")
+            return false
+        }
```
Comment (emdobrin) [inline AEPIdentity/Sources/IdentityState.swift:49]: can be reverted, as you long inside readyForSyncIdentifiers

File: AEPIdentity/Sources/IdentityState.swift [hunk 4]
```diff
@@ -349,7 +355,12 @@ class IdentityState {
     ///   - event: event responsible for the hit
     ///   - addConsentFlag: flag indicating if the adId changed
     private func queueHit(identityProperties: IdentityProperties, configSharedState: [String: Any], event: Event, addConsentFlag: Bool) {
-        let server = configSharedState[IdentityConstants.Configuration.EXPERIENCE_CLOUD_SERVER] as? String ?? IdentityConstants.Default.SERVER
+        var server = configSharedState[IdentityConstants.Configuration.EXPERIENCE_CLOUD_SERVER] as? String ?? ""
+
+        if server.isEmpty {
+            Log.trace(label: "\(LOG_TAG):\(#function)", "Setting to default server url ( \(IdentityConstants.Default.SERVER)), \(IdentityConstants.Configuration.EXPERIENCE_CLOUD_SERVER) not found in configuration.")
```
Comment (emdobrin) [inline AEPIdentity/Sources/IdentityState.swift:361]: We probably don't need this log, this is an expected scenario.

**Review by emdobrin — COMMENTED** [review]
Looks good, with some small comments regarding logging

File: AEPIdentity/Tests/IdentityTests.swift [hunk 5]
```diff
@@ -307,6 +307,81 @@ class IdentityTests: XCTestCase {
...
+        mockRuntime.simulateSharedState(extensionName: IdentityConstants.SharedStateKeys.CONFIGURATION, event: event, data: (configData, .set))
+        let _ = identity.readyForEvent(event)
+
+        // test
+        mockRuntime.simulateComingEvent(event: event)
+
+        // verify network request is sent
+        let mockNetworkService = ServiceProvider.shared.networkService as! MockNetworkServiceOverrider
+        let optOutHitHost = mockNetworkService.connectAsyncCalledWithNetworkRequest?.url.host // network request for opt-out hit should have been sent
+        XCTAssertTrue(optOutHitHost!.contains("dpm.demdex.net"))
```
Comment (kevinlind) [inline AEPIdentity/Tests/IdentityTests.swift:332]: Let's unwrap `optOutHitHost` so this test won't crash in the case where the network call doesn't actually get sent.  Maybe just use a guard and if the host is nil then `XCTFail()` the test?

File: AEPIdentity/Tests/IdentityTests.swift [hunk 6]
```diff
@@ -307,6 +307,81 @@ class IdentityTests: XCTestCase {
...
```
Comment (kevinlind) [inline AEPIdentity/Tests/IdentityTests.swift:357]: Same here about unwrapping `optOutHitHost`.

File: AEPIdentity/Tests/IdentityTests.swift [hunk 7]
```diff
@@ -307,6 +307,81 @@ class IdentityTests: XCTestCase {
...
+        mockRuntime.simulateSharedState(extensionName: IdentityConstants.SharedStateKeys.CONFIGURATION, event: event, data: (configData, .set))
+        let _ = identity.readyForEvent(event)
+
+        // test
+        mockRuntime.simulateComingEvent(event: event)
+
+        // verify network request is sent
+        let mockNetworkService = ServiceProvider.shared.networkService as! MockNetworkServiceOverrider
+        let optOutHitHost = mockNetworkService.connectAsyncCalledWithNetworkRequest?.url.host // network request for opt-out hit should have been sent
+        XCTAssertTrue(optOutHitHost!.contains("example.com"))
```
Comment (kevinlind) [inline AEPIdentity/Tests/IdentityTests.swift:382]: Same here about unwrapping `optOutHitHost`.

**Review by kevinlind — COMMENTED** [review]
In `readyForEvent`, should we add a trace log of the Event so we know when each event is being processed?

**Review by emdobrin — APPROVED** [review]

**Review by kevinlind — APPROVED** [review]

**Pull request merged**